src-java/reroute-topology/reroute-storm-topology/src/main/java/org/openkilda/wfm/topology/reroute/service/OperationQueueService.java

```diff
@@ -0,0 +1,135 @@
+
+    public OperationQueueService(OperationQueueCarrier carrier) {
+        this.carrier = carrier;
+    }
+
+    /**
+     * Add command to the front of the queue.
+     */
+    public void addFirst(String flowId, String correlationId, CommandData data) {
+        log.info("Flow command {} has been added to the queue for flow {} with correlation id {}",
+                data, flowId, correlationId);
+        FlowQueueData queueData = getFlowQueueData(flowId);
+        queueData.getQueue().addFirst(new OperationData(correlationId, data));
+        processQueue(queueData);
+    }
+
+    /**
+     * Add command to the end of the queue.
+     */
+    public void addLast(String flowId, String correlationId, CommandData data) {
+        log.info("Flow command {} has been added to the queue for flow {} with correlation id {}",
+                data, flowId, correlationId);
+        FlowQueueData queueData = getFlowQueueData(flowId);
+        queueData.getQueue().addLast(new OperationData(correlationId, data));
+        processQueue(queueData);
+    }
+
+    /**
+     * Handle timeout.
+     */
+    public void handleTimeout(String correlationId) {
+        log.warn("Operation with correlationId {} has timed out", correlationId);
+        Collection<FlowQueueData> foundQueueData = flowCommands.values().stream()
+                .filter(flowQueueData -> correlationId.equals(flowQueueData.getTaskInProgress()))
+                .collect(Collectors.toList());
+
+        if (foundQueueData.isEmpty()) {
+            log.warn("No queue data with correlationId {} found. Timeout event skipped.", correlationId);
+        } else if (foundQueueData.size() > 1) {
+            log.error("Found more than one queue data with correlationId {}. Timed out all of them.", correlationId);
+        }
+
+        foundQueueData.forEach(queueData -> operationCompleted(queueData, null));
+    }
+
+    public void operationCompleted(String flowId, InfoData response) {
+        operationCompleted(getFlowQueueData(flowId), response);
+    }
+
+    private void operationCompleted(FlowQueueData queueData, InfoData response) {
+        log.info("Flow command {} has been completed with response {}", queueData.getTaskInProgress(), response);
+        queueData.setTaskInProgress(null);
+        processQueue(queueData);
+    }
+
+    private void processQueue(FlowQueueData queueData) {
+        if (!queueData.getQueue().isEmpty() && !queueData.isOperationInProgress()) {
+            OperationData operationData = queueData.getQueue().pop();
+            carrier.emitRequest(operationData.getCorrelationId(), operationData.getCommandData());
+            queueData.setTaskInProgress(operationData.getCorrelationId());
+            log.info("Flow command {} has been sent", operationData.getCommandData());
+        }
+    }
+
+    private FlowQueueData getFlowQueueData(String flowId) {
+        return flowCommands.computeIfAbsent(flowId, v -> new FlowQueueData());
```

Comment:
Have we ever clean up this map? This approach creates a memory leak that can be triggered/used via NB API (you need to produce reroute requests for not existing/unique flowId... endlessly).

Reply:
Good point! Clean up has been added.